File: languages/php/cross-language-features.php
<?php
require_once __DIR__ . '/../../runtimes/php/willyhorizont/runtime.php';
use Runtimes\Php\WillyHorizont\Runtime;

/*
1. support closure as value, or has workaround
*/
$say_hello = function ($callback_function) {
    echo "hello" . PHP_EOL;
    $callback_function();
};
$say_hello(function () {
    echo "world" . PHP_EOL;
});
$create_multiplier = fn ($aa) => fn ($bb) => ($aa * $bb);
$multiply_by_two = $create_multiplier(2);
echo "multiply_by_two(10): " . $multiply_by_two(10) . PHP_EOL;
$multiply_by_eight = $create_multiplier(8);
echo "multiply_by_eight(4): " . $multiply_by_eight(4) . PHP_EOL;
echo "multiply_by_two(8): " . $multiply_by_two(8) . PHP_EOL;

/*
2. support dynamic-typed value, or has workaround
*/
$xl_list = [
    null,
    true,
    false,
    "foo",
    0,
    -123,
    123.789,
    -123.789,
    [1, 2, 3],
    ["foo" => "bar"],
    fn ($aa, $bb) => ($aa * $bb),
];
echo "xl_list" . Runtime::json_stringify($xl_list) . PHP_EOL;
echo "xl_list" . Runtime::json_stringify($xl_list, ["pretty" => true]) . PHP_EOL;
$xl_dict = [
    "xl_none" => null,
    "xl_bool_true" => true,
    "xl_bool_false" => false,
    "xl_string" => "foo",
    "xl_int_positive" => 0,
    "xl_int_negative" => -123,
    "xl_float_positive" => 123.789,
    "xl_float_negative" => -123.789,
    "xl_list" => [1, 2, 3],
    "xl_dict" => ["foo" => "bar"],
    "xl_closure" => fn ($aa, $bb) => ($aa * $bb),
];
echo "xl_dict" . Runtime::json_stringify($xl_dict) . PHP_EOL;
echo "xl_dict" . Runtime::json_stringify($xl_dict, ["pretty" => true]) . PHP_EOL;

File: languages/python/cross-language-features.py
from runtimes.python.willyhorizont.runtime import json_stringify

# 1. support closure as value, or has workaround
say_hello = lambda callback_function: [
    print("hello"),
    callback_function(),
][-1]
say_hello(lambda: print("wold"))
create_multiplier = lambda aa: lambda bb: (aa * bb)
multiply_by_two = create_multiplier(2)
print(f"multiply_by_two(10): {multiply_by_two(10)}")
multiply_by_eight = create_multiplier(8)
print(f"multiply_by_eight(4): {multiply_by_eight(4)}")
print(f"multiply_by_two(8): {multiply_by_two(8)}")

# 2. support dynamic-typed value, or has workaround
xl_list = [
    None,
    True,
    False,
    "foo",
    0,
    -123,
    123.789,
    -123.789,
    [1, 2, 3],
    {"foo":"bar"},
    lambda aa, bb: (aa * bb),
]
print(f"xl_list: {json_stringify(xl_list)}")
print(f"xl_list: {json_stringify(xl_list, pretty=True)}")
xl_dict = {
    "xl_none": None,
    "xl_bool_true": True,
    "xl_bool_false": False,
    "xl_string": "foo",
    "xl_int_positive": 0,
    "xl_int_negative": -123,
    "xl_float_positive": 123.789,
    "xl_float_negative": -123.789,
    "xl_list": [1, 2, 3],
    "xl_dict": {"foo":"bar"},
    "xl_closure": lambda aa, bb: (aa * bb),
}
print(f"xl_dict: {json_stringify(xl_dict)}")
print(f"xl_dict: {json_stringify(xl_dict, pretty=True)}")
